src/D2L.CodeStyle.Analyzers/Async/Generator/FileCollector.cs

```diff
@@ -0,0 +1,304 @@
+				// for "record struct" etc.
+				keyword2: secondaryKeyword,
+				identifier: decl.Identifier,
+				typeParameterList: decl.TypeParameterList,
+				constraintClauses: decl.ConstraintClauses,
+				openBrace: decl.OpenBraceToken
+			);
+
+			var wroteSomethingInChild = WriteChildren( decl );
+			var wroteMethodsForMe = WriteMethods( decl );
+
+			var wroteSomething = wroteSomethingInChild || wroteMethodsForMe;
+
+			CloseBraceIfNecessary( wroteSomething, decl.CloseBraceToken );
+
+			CheckPremableInvariant( wroteSomething, preambleCountBefore );
+
+			return wroteSomething;
+		}
+
+		private void AddPartialTypeToPreamble(
+			SyntaxNode decl,
+			SyntaxToken keyword1,
+			SyntaxToken keyword2,
+			SyntaxToken identifier,
+			TypeParameterListSyntax? typeParameterList,
+			SyntaxList<TypeParameterConstraintClauseSyntax> constraintClauses,
+			SyntaxToken openBrace
+		) {
+			// Create the partial keyword and grab all leading trivia from the node
+			var partialKeyword = CreatePartialKeyword( decl );
+
+			// Clear any leading trivia from the keyword, if it exists.
+			// Only the first token on the line should have leading trivia, and
+			// we've moved that onto partialKeyword.
+			// This could get ugly if someone did
+			//
+			//   sealed class
+			//      Foo {
+			keyword1 = keyword1.WithLeadingTrivia( EmptyTrivia );
+			keyword2 = keyword2.WithLeadingTrivia( EmptyTrivia );
+			identifier = identifier.WithLeadingTrivia( EmptyTrivia );
+
+			m_preambles.AddLast(
+				new PieceOfSyntax[] {
+					partialKeyword,
+					keyword1,
+					keyword2,
+					identifier,
+					(typeParameterList != null ? typeParameterList : NoneToken ),
+					openBrace
+				}
+			);
+		}
+
+		private static SyntaxToken CreatePartialKeyword( SyntaxNode decl )
+			=> SyntaxFactory.Token( SyntaxKind.PartialKeyword )
+				.WithLeadingTrivia( decl.GetLeadingTrivia() )
+				.WithTrailingTrivia( SyntaxFactory.Space );
+
+		private void CloseBraceIfNecessary( bool wroteSomething, SyntaxToken? closeBrace ) {
+			if( wroteSomething ) {
+				if( closeBrace.HasValue ) {
+					m_out.Append( closeBrace.Value.ToFullString() );
+				}
+			} else {
+				// Throw away the start of the decl that we never wrote
+				m_preambles.RemoveLast();
+			}
+		}
+
+		private bool WriteMethods( SyntaxNode parent ) {
+			if( !m_methods.TryGetValue( parent, out var methods ) ) {
+				return false;
+			}
+
+			m_methods.Remove( parent );
+
+			// Flush any preambles in the queue
+			while( m_preambles.Count != 0 ) {
+				foreach( var thing in m_preambles.First() ) {
+					m_out.Append( thing );
+				}
+				m_preambles.RemoveFirst();
+			}
+
+			foreach( var method in methods ) {
+				m_out.Append( method );
+			}
+
+			return true;
+		}
+
+		// Inline assertions to catch some obvious bugs quickly
+		private void CheckPremableInvariant( bool wroteSomething, int countBefore ) {
+			if( wroteSomething && m_preambles.Count != 0 ) {
+				throw new BugException( "Left over preambles" );
+			}

```

Comment:
Will this stuff work for sibling classes?

```
namespace Foo {

   class X {}

   class Y {}
   
```

Reply:
Yep. So when you are in Foo then countBefore will be 1, and the idea is that after recursing into Foo the only valid values should be 1 (i.e. we did nothing inside X, so we pushed but then discarded `X`s preamble) or 0 (because we flushed them all to write out a method). Loop invariant basically.

Reply:
This should be covered by tests I think, although i was a bit unstructured with it.

Reply:
K, I'm not sure if I'm following it through the logic properly.

If I'm recursed in such that my base preamble count is `1`, then how is it `0` between children?

```csharp
			bool wroteSomething = false;

			// First recurse into child types
			foreach( var child in node.ChildNodes() ) {
				bool wroteSomethingHere = child switch {
					ClassDeclarationSyntax @class => WriteType( @class ),
					StructDeclarationSyntax @struct => WriteType( @struct ),
					InterfaceDeclarationSyntax @interface => WriteType( @interface ),

					// Covers record + record struct:
					RecordDeclarationSyntax @record => WriteType( @record, @record.ClassOrStructKeyword ),

					// Covers file-scoped and block namespaces
					BaseNamespaceDeclarationSyntax @namespace => WriteNamespace( @namespace ),

					// Ignore any other child node
					_ => false
				};

				if( wroteSomethingHere ) {
					wroteSomething = true;
				}

				CheckPremableInvariant( wroteSomething, preambleCountBefore );
			}
```

You're right that the "Complicated" test appears to cover it though.

Reply:
Oh okay right, because preambles goes to 0 when we wrote all of the namespace information, we're not rewriting each time.

K